### PR TITLE
[IOTDB-527] Refactor series reader

### PR DESCRIPTION
**Pull request by qiaojialin**
remove hasOverlappedPage & isPageOverlapped methods in IAggregateReader

optimize logic in series reader

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/query/reader/series/SeriesReader.java [hunk 1]
```diff
@@ -104,251 +130,274 @@ public SeriesReader(Path seriesPath, TSDataType dataType, QueryContext context,
 
 
   public boolean hasNextChunk() throws IOException {
-    if (hasCachedFirstChunkMetadata) {
-      return true;
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/reader/series/SeriesReader.java]: It seems that if I keep calling hasNextChunk(), the firstChunkMetaData will move to next all the time.
you can add an if statement.
if (firstChunkMetaData != null)
    return true;

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java [hunk 2]
```diff
@@ -114,159 +105,202 @@ protected RowRecord nextWithoutConstraint() throws IOException {
...
-      BatchData lastBatch = cachedBatchDataList.get(index);
-
-      calcBatchData(result, lastBatch);
-      if (isEndCalc(result, lastBatch)) {
-        isCalculatedList.add(true);
-        remainingToCalculate--;
-        if (remainingToCalculate == 0) {
-          return aggregateResultList;
+
+  private class GroupByExecutor {
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java]: It's better name it GroupByTask

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java [hunk 3]
```diff
@@ -64,45 +67,33 @@ public GroupByWithoutValueFilterDataSet(QueryContext context, GroupByPlan groupB
...
-
-        IAggregateReader seriesReader = new SeriesAggregateReader(path, dataTypes.get(i), context,
-            queryDataSource, timeFilter, null, null);
-        aggregateReaders.put(path, seriesReader);
-      }
+
+      QueryDataSource queryDataSource = QueryResourceManager.getInstance()
+          .getQueryDataSource(path, context, timeFilter);
+      // update filter by TTL
+      timeFilter = queryDataSource.updateFilterUsingTTL(timeFilter);
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java]: No need to construct these two, if pathExecutors already has the path

Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java]: done

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java (hunk 2 above)
Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java]: no , i like executor

**Comment by liutaohua** [comment]
> About the Group by parallel, it seems that it will create new future task and invoke future.get() each time nextWithoutConstraint() is called. This way is like the first version's raw data query parallel that I implements. It will slower the query instead of speeding up.

Thank you very much for your reminding. I will continue to study why is it slow

File: server/src/main/java/org/apache/iotdb/db/query/reader/series/SeriesReaderByTimestamp.java [hunk 4]
```diff
@@ -63,24 +61,39 @@ public SeriesReaderByTimestamp(Path seriesPath, TSDataType dataType, QueryContex
...
+      return true;
+    }
+
+    /*
+     * consume chunk secondly
+     */
     while (seriesReader.hasNextChunk()) {
-      if (!satisfyTimeFilter(seriesReader.currentChunkStatistics())) {
-        seriesReader.skipCurrentChunk();
-        continue;
```
[the remark this answers is not in the document]
Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/reader/series/SeriesReaderByTimestamp.java]: I've made up for it

File: server/src/main/java/org/apache/iotdb/db/query/reader/series/SeriesReader.java (hunk 1 above)
Comment (qiaojialin) [reply server/src/main/java/org/apache/iotdb/db/query/reader/series/SeriesReader.java]: fixed

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java (hunk 2 above)
Comment (qiaojialin) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java]: executor or task are all fine to me

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBCloseIT.java [hunk 5]
```diff
@@ -85,7 +86,7 @@ private static void insertData()
             .getConnection(Config.IOTDB_URL_PREFIX + "127.0.0.1:6667/", "root", "root");
          Statement statement = connection.createStatement()) {
 
-      for (String sql : Constant.create_sql) {
+      for (String sql : org.apache.iotdb.db.constant.TestConstant.create_sql) {
```
Comment (samperson1997) [inline server/src/test/java/org/apache/iotdb/db/integration/IoTDBCloseIT.java]:
```suggestion
      for (String sql : TestConstant.create_sql) {
```
And all the same appearance in this test class : )

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBEngineTimeGeneratorIT.java [hunk 6]
```diff
@@ -100,7 +101,7 @@ private static void insertData() throws ClassNotFoundException, SQLException {
         Statement statement = connection.createStatement()) {
 
       // create storage group and measurement
-      for (String sql : Constant.create_sql) {
+      for (String sql : org.apache.iotdb.db.constant.TestConstant.create_sql) {
```
Comment (samperson1997) [inline server/src/test/java/org/apache/iotdb/db/integration/IoTDBEngineTimeGeneratorIT.java]:
```suggestion
      for (String sql : TestConstant.create_sql) {
```
And all the same appearance in this test class : )

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBLargeDataIT.java [hunk 7]
```diff
@@ -89,7 +90,7 @@ private static void insertData()
         .getConnection(Config.IOTDB_URL_PREFIX + "127.0.0.1:6667/", "root", "root");
         Statement statement = connection.createStatement()) {
 
-      for (String sql : Constant.create_sql) {
+      for (String sql : org.apache.iotdb.db.constant.TestConstant.create_sql) {
```
Comment (samperson1997) [inline server/src/test/java/org/apache/iotdb/db/integration/IoTDBLargeDataIT.java]:
```suggestion
      for (String sql : TestConstant.create_sql) {
```
And all the same appearance in this test class : )

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBMultiSeriesIT.java [hunk 8]
```diff
@@ -43,7 +44,7 @@
  */
 public class IoTDBMultiSeriesIT {
 
-  private static boolean testFlag = Constant.testFlag;
+  private static boolean testFlag = org.apache.iotdb.db.constant.TestConstant.testFlag;
```
Comment (samperson1997) [inline server/src/test/java/org/apache/iotdb/db/integration/IoTDBMultiSeriesIT.java]:
```suggestion
  private static boolean testFlag = TestConstant.testFlag;
```
And all the same appearance in this test class : )

Comment (samperson1997) [reply server/src/test/java/org/apache/iotdb/db/integration/IoTDBMultiSeriesIT.java]: And also in other test classes...

Comment (qiaojialin) [reply server/src/test/java/org/apache/iotdb/db/integration/IoTDBMultiSeriesIT.java]: fixed

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBLargeDataIT.java (hunk 7 above)
Comment (qiaojialin) [reply server/src/test/java/org/apache/iotdb/db/integration/IoTDBLargeDataIT.java]: fixed

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBEngineTimeGeneratorIT.java (hunk 6 above)
Comment (qiaojialin) [reply server/src/test/java/org/apache/iotdb/db/integration/IoTDBEngineTimeGeneratorIT.java]: fixed

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBCloseIT.java (hunk 5 above)
Comment (qiaojialin) [reply server/src/test/java/org/apache/iotdb/db/integration/IoTDBCloseIT.java]: fixed

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java [hunk 9]
```diff
@@ -114,159 +97,194 @@ protected RowRecord nextWithoutConstraint() throws IOException {
...
+      for (Entry<Path, GroupByExecutor> pathGroupByExecutorEntry : pathExecutors.entrySet()) {
+        GroupByExecutor executor = pathGroupByExecutorEntry.getValue();
+        executor.resetAggregateResults();
+        List<Pair<AggregateResult, Integer>> aggregations = executor.calcResult();
+        for (int i = 0; i < aggregations.size(); i++) {
+          fields[aggregations.get(i).right] = aggregations.get(i).left;
+        }
       }
+    } catch (QueryProcessException e) {
+      logger.error("GroupByWithoutValueFilterDataSet execute has error,{}", e);
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java]: no need to use {},just
logger.error("GroupByWithoutValueFilterDataSet execute has error ", e);

File: server/src/main/java/org/apache/iotdb/db/query/reader/series/SeriesReader.java [hunk 10]
```diff
@@ -104,251 +126,305 @@ public SeriesReader(Path seriesPath, TSDataType dataType, QueryContext context,
...
+      /*
+       * first chunk metadata is already unpacked, consume cached pages
+       */
+      if (!cachedPageReaders.isEmpty()) {
+        firstPageReader = cachedPageReaders.poll();
+      }
+    }
+
+    if (firstPageReader != null && !cachedPageReaders.isEmpty() &&
+        firstPageReader.data.getStatistics().getEndTime() >=
```
[the remark this answers is not in the document]
Comment (qiaojialin) [reply server/src/main/java/org/apache/iotdb/db/query/reader/series/SeriesReader.java]: fixed

File: server/src/main/java/org/apache/iotdb/db/query/reader/series/SeriesReader.java [hunk 11]
```diff
@@ -55,26 +55,48 @@
...
       new PriorityQueue<>(Comparator.comparingLong(ChunkMetaData::getStartTime));
 
-  private boolean hasCachedFirstChunkMetadata;
-  private ChunkMetaData firstChunkMetaData;
-
-  private PriorityQueue<VersionPair<IPageReader>> overlappedPageReaders =
+  /*
+   * page cache
+   */
+  private VersionPair<IPageReader> firstPageReader;
```
[the remark this answers is not in the document]
Comment (qiaojialin) [reply server/src/main/java/org/apache/iotdb/db/query/reader/series/SeriesReader.java]: fixed

File: server/src/main/java/org/apache/iotdb/db/query/reader/series/SeriesReader.java [hunk 12]
```diff
@@ -55,26 +55,48 @@
...
+  private PriorityQueue<VersionPair<IPageReader>> cachedPageReaders =
```
[the remark this answers is not in the document]
Comment (qiaojialin) [reply server/src/main/java/org/apache/iotdb/db/query/reader/series/SeriesReader.java]: fixed

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java [hunk 13]
```diff
@@ -64,45 +63,29 @@ public GroupByWithoutValueFilterDataSet(QueryContext context, GroupByPlan groupB
...
-        QueryDataSource queryDataSource = QueryResourceManager.getInstance()
-            .getQueryDataSource(path, context, timeFilter);
-        // update filter by TTL
-        timeFilter = queryDataSource.updateFilterUsingTTL(timeFilter);
-
-        IAggregateReader seriesReader = new SeriesAggregateReader(path, dataTypes.get(i), context,
-            queryDataSource, timeFilter, null, null);
-        aggregateReaders.put(path, seriesReader);
+      if (!pathExecutors.containsKey(path)) {
+        //init reader
```
[the remark this answers is not in the document]
Comment (qiaojialin) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java]: fixed

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java [hunk 14]
```diff
@@ -114,159 +97,194 @@ protected RowRecord nextWithoutConstraint() throws IOException {
...
-        if (remainingToCalculate == 0) {
-          return aggregateResultList;
+  private class GroupByExecutor {
+
+    private IAggregateReader reader;
+    private BatchData preCachedData;
+    //<aggFunction - indexForRecord> of path
+    private List<Pair<AggregateResult, Integer>> results = new ArrayList<>();
+
+    public GroupByExecutor(Path path, TSDataType dataType, QueryContext context, Filter timeFilter)
```
[the remark this answers is not in the document]
Comment (qiaojialin) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java]: fixed

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBMultiStatementsIT.java [hunk 15]
```diff
@@ -85,7 +86,7 @@ private static void insertData()
             .getConnection(Config.IOTDB_URL_PREFIX + "127.0.0.1:6667/", "root", "root");
          Statement statement = connection.createStatement()) {
 
-      for (String sql : Constant.create_sql) {
+      for (String sql : org.apache.iotdb.db.constant.TestConstant.create_sql) {
```
Comment (JackieTien97) [inline server/src/test/java/org/apache/iotdb/db/integration/IoTDBMultiStatementsIT.java]:
```suggestion
      for (String sql : TestConstant.create_sql) {
```

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBMultiStatementsIT.java [hunk 16]
```diff
@@ -138,7 +139,7 @@ public void selectAllTest() throws ClassNotFoundException {
       int cnt1 = 0;
       while (resultSet1.next() && cnt1 < 5) {
         StringBuilder builder = new StringBuilder();
-        builder.append(resultSet1.getString(Constant.TIMESTAMP_STR))
+        builder.append(resultSet1.getString(org.apache.iotdb.db.constant.TestConstant.TIMESTAMP_STR))
```
Comment (JackieTien97) [inline server/src/test/java/org/apache/iotdb/db/integration/IoTDBMultiStatementsIT.java]:
```suggestion
        builder.append(resultSet1.getString(TestConstant.TIMESTAMP_STR))
```

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBMultiStatementsIT.java [hunk 17]
```diff
@@ -154,7 +155,7 @@ public void selectAllTest() throws ClassNotFoundException {
       int cnt2 = 0;
       while (resultSet2.next()) {
         StringBuilder builder = new StringBuilder();
-        builder.append(resultSet2.getString(Constant.TIMESTAMP_STR))
+        builder.append(resultSet2.getString(org.apache.iotdb.db.constant.TestConstant.TIMESTAMP_STR))
```
Comment (JackieTien97) [inline server/src/test/java/org/apache/iotdb/db/integration/IoTDBMultiStatementsIT.java]:
```suggestion
        builder.append(resultSet2.getString(TestConstant.TIMESTAMP_STR))
```

File: server/src/test/java/org/apache/iotdb/db/tools/IoTDBWatermarkTest.java [hunk 18]
```diff
@@ -48,8 +45,8 @@
  */
 public class IoTDBWatermarkTest {
 
-  private static String filePath1 = TestConstant.BASE_OUTPUT_PATH.concat("watermarked_query_result.csv");
-  private static String filePath2 = TestConstant.BASE_OUTPUT_PATH.concat("notWatermarked_query_result.csv");
+  private static String filePath1 = org.apache.iotdb.db.constant.TestConstant.BASE_OUTPUT_PATH.concat("watermarked_query_result.csv");
+  private static String filePath2 = org.apache.iotdb.db.constant.TestConstant.BASE_OUTPUT_PATH.concat("notWatermarked_query_result.csv");
```
Comment (JackieTien97) [inline server/src/test/java/org/apache/iotdb/db/tools/IoTDBWatermarkTest.java]:
```suggestion
  private static String filePath1 = TestConstant.BASE_OUTPUT_PATH.concat("watermarked_query_result.csv");
  private static String filePath2 = TestConstant.BASE_OUTPUT_PATH.concat("notWatermarked_query_result.csv");
```

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBMultiStatementsIT.java (hunk 15 above)
Comment (qiaojialin) [reply server/src/test/java/org/apache/iotdb/db/integration/IoTDBMultiStatementsIT.java]: fixed

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java (hunk 9 above)
Comment (qiaojialin) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java]: fixed

**Pull request merged**